File: bfs.js
class Queue {
  constructor() {
    this.items = [];
  }

  enqueue(element) {
    this.items.push(element);
  }

  dequeue() {
    if (this.isEmpty()) {
      return;
    }
    return this.items.shift();
  }

  isEmpty() {
    return this.items.length === 0;
  }
}

let graph = new Map();
graph.set("1", ["2", "3", "4"]);
graph.set("2", []);
graph.set("3", ["5", "6"]);
graph.set("4", []);
graph.set("5", []);
graph.set("6", []);

const bfs = (graph, start) => {
  let count = 0;
  let queue = new Queue();
  let visited = new Set();
  let arr = [];

  const bfsRecursive = () => {
    if (queue.isEmpty()) {
      return;
    }

    let vertex = queue.dequeue();
    if (!visited.has(vertex)) {
      visited.add(vertex);
      arr[count++] = vertex;
      console.log(`visited vertex : ${vertex}, count: ${count}`);
      let neighbors = graph.get(vertex);
      for (let neighbor of neighbors) {
        if (!visited.has(neighbor)) {
          queue.enqueue(neighbor);
        }
      }
    }
    bfsRecursive();
  };

  queue.enqueue(start);
  bfsRecursive();
  console.log(`BFS order : ${arr.join(" ")}`);
};

bfs(graph, "1");

//interaktif
const readline = require("readline");

class Queue2 {
  constructor() {
    this.items = [];
  }

  enqueue(element) {
    this.items.push(element);
  }

  dequeue() {
    if (this.isEmpty()) {
      return;
    }
    return this.items.shift();
  }

  isEmpty() {
    return this.items.length === 0;
  }
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let graph2 = new Map();

const askGraphInput = () => {
  rl.question(
    'Masukkan edge (node1 node2, atau "done" untuk selesai): ',
    (input) => {
      if (input.toLowerCase() === "done") {
        askStartNode();
      } else {
        const [node1, node2] = input.split(" ");
        if (!graph2.has(node1)) {
          graph2.set(node1, []);
        }
        if (!graph2.has(node2)) {
          graph2.set(node2, []);
        }
        graph2.get(node1).push(node2);
        graph2.get(node2).push(node1);
        askGraphInput();
      }
    }
  );
};

const askStartNode = () => {
  rl.question("Masukkan node edge awal BFS: ", (start) => {
    bfs2(graph2, start);
    rl.close();
  });
};

const bfs2 = (graph2, start) => {
  let queue = new Queue2();
  let visited = new Set();
  let count = 0;
  let arr = [];

  const bfsRecursive = () => {
    if (queue.isEmpty()) {
      return;
    }

    let vertex = queue.dequeue();
    if (!visited.has(vertex)) {
      visited.add(vertex);
      arr[count++] = vertex;
      console.log(`visited vertex : ${vertex}, count: ${count}`);

      let neighbors = graph2.get(vertex);
      for (let neighbor of neighbors) {
        if (!visited.has(neighbor)) {
          queue.enqueue(neighbor);
        }
      }
    }
    bfsRecursive();
  };

  queue.enqueue(start);
  bfsRecursive();
  console.log(`BFS order : ${arr.join(" ")}`);
};

console.log('Input edge graf (misal: "A B"). Ketik "done" untuk keluar.');
askGraphInput();

File: projek/src/bfs.java
import java.util.*;

class Queue {
    private LinkedList<Integer> items;

    public Queue() {
        items = new LinkedList<>();
    }

    public void enqueue(int element) {
        items.addLast(element);
    }

    public int dequeue() {
        if (isEmpty()) {
            return -1;
        }
        return items.removeFirst();
    }

    public boolean isEmpty() {
        return items.isEmpty();
    }
}

public class bfs {
    private static Map<Integer, List<Integer>> graph = new HashMap<>();
    private static Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) {
        System.out.println("Input edge graf (misal: \"1 2\"). Ketik \"done\" untuk selesai.");
        askGraphInput();
    }

    private static void askGraphInput() {
        while (true) {
            System.out.print("Masukkan edge (node1 node2, atau \"done\" untuk selesai): ");
            String input = scanner.nextLine();
            if (input.equalsIgnoreCase("done")) {
                askStartNode();
                break;
            } else {
                String[] nodes = input.split(" ");
                int node1 = Integer.parseInt(nodes[0]);
                int node2 = Integer.parseInt(nodes[1]);

                graph.putIfAbsent(node1, new LinkedList<>());
                graph.putIfAbsent(node2, new LinkedList<>());

                graph.get(node1).add(node2);
                graph.get(node2).add(node1);
            }
        }
    }

    private static void askStartNode() {
        System.out.print("Masukkan node awal BFS: ");
        int start = scanner.nextInt();
        bfs(start);
    }

    private static void bfs(int start) {
        Queue queue = new Queue();
        Set<Integer> visited = new HashSet<>();
        List<Integer> order = new ArrayList<>();
        int count = 0;

        queue.enqueue(start);

        while (!queue.isEmpty()) {
            int vertex = queue.dequeue();
            if (!visited.contains(vertex)) {
                visited.add(vertex);
                order.add(vertex);
                count++;
                System.out.println("visited vertex : " + vertex + ", count: " + count);

                List<Integer> neighbors = graph.get(vertex);
                if (neighbors != null) {
                    for (int neighbor : neighbors) {
                        if (!visited.contains(neighbor)) {
                            queue.enqueue(neighbor);
                        }
                    }
                }
            }
        }

        System.out.println("BFS order: " + order);
    }
}
